### ISSUE TITLE
<system_error>: std::error_code() does not compare equal to std::error_condition()

### ISSUE THREAD

**Issue by pdimov** [microsoft/STL#2213]
As the title says, really.
```
#include <system_error>
#include <cstdio>

int main()
{
    if( std::error_code() == std::error_condition() )
    {
        std::puts( "equal" );
    }
    else
    {
        std::puts( "not equal" );
    }
}
```
prints "not equal", but I would expect these two to compare equal, as both represent the same "no error" condition.

**Comment by fsb4000** [comment]
gcc and clang says different: https://gcc.godbolt.org/z/faMoPz3bj

**Comment by CaseyCarter** [comment]
As @fsb4000 points out, `error_code() == error_condition()` reduces to "`0` corresponds to a POSIX `errno` value" where "What constitutes correspondence for any given operating system is unspecified" ([[syserr.errcat.objects]/4](https://eel.is/c++draft/syserr.errcat.objects#4)). To my knowledge `0` is generally a success code on POSIX/Win32 which consequently does not correspond to any POSIX `errno` value, in which case I'd expect `error_code() != error_condition()` to hold as it does for MSVCSTL and libstdc++. 

I'm no Darwin expert, but perhaps `0` is a common error code for OS APIs there, justifying libc++'s choice? Googling "Darwin errno.h" turns up a header (https://opensource.apple.com/source/xnu/xnu-2050.7.9/bsd/sys/errno.h) that notably doesn't define an error for `0`, however. In any case, the "correspondence ... is unspecified" escape clause seems to allow implementors complete freedom here.

TLDR: I don't think there's a bug here, it doesn't seem to be WG21's intent that `error_code() == error_condition()` is `true`.

**Comment by StephanTLavavej** [comment]
This is an issue in the Standard, now tracked by LWG-3598 "`system_category().default_error_condition(0)` is underspecified". We expect that when this is resolved in the Working Paper (at which point we'll file a GitHub issue here labeled `LWG`), our implementation will need to change accordingly. We hope that this will be possible without breaking binary compatibility (there's separately compiled code in `syserror.cpp`, but we should be able to make a header-only change that bypasses it for this case).

Thanks for reporting this! :smile_cat:

**Comment by fsb4000** [comment]
@pdimov Hi. We implemented the change and it is expected in VS 2022 17.2 Preview 3

**Comment by pdimov** [comment]
Thanks!
